lib/spack/spack/patch.py

```diff
@@ -58,76 +59,118 @@ def apply_patch(stage, patch_path, level=1, working_dir="."):
         self.owner = pkg.fullname
         self.path_or_url = path_or_url  # needed for debug output
-        self.path = None  # must be set before apply()
+        self.path: Optional[str] = None  # must be set before apply()
         self.level = level
         self.working_dir = working_dir
+        self.sha256 = sha256
 
-    def apply(self, stage: "spack.stage.Stage"):
+    def apply(self, stage: spack.stage.Stage) -> None:
         """Apply a patch to source in a stage.
 
-        Arguments:
-            stage (spack.stage.Stage): stage where source code lives
+        Args:
+            stage: stage where source code lives
         """
         if not self.path or not os.path.isfile(self.path):
             raise NoSuchPatchError(f"No such patch: {self.path}")
 
         apply_patch(stage, self.path, self.level, self.working_dir)
 
-    @property
-    def stage(self):
-        return None
+    # TODO: Use TypedDict once Spack supports Python 3.8+ only
+    def to_dict(self) -> Dict[str, Any]:
+        """Dictionary representation of the patch.
 
-    def to_dict(self):
-        """Partial dictionary -- subclases should add to this."""
+        Returns:
+            A dictionary representation.
+        """
         return {
             "owner": self.owner,
             "sha256": self.sha256,
             "level": self.level,
             "working_dir": self.working_dir,
         }
 
-    def __eq__(self, other):
+    def __eq__(self, other: object) -> bool:
+        """Equality check.
+
+        Args:
+            other: another patch
+
+        Returns:
+            True if both patches have the same checksum, else False
+
+        Raises:
+            AssertionError: If *other* is not of type Patch.
+        """
+        assert hasattr(other, "sha256"), "*other* must be of type Patch"
         return self.sha256 == other.sha256
 
-    def __hash__(self):
+    def __hash__(self) -> int:
+        """Unique hash.
+
+        Returns:
+            A unique hash based on the sha256.
+        """
         return hash(self.sha256)
 
 
 class FilePatch(Patch):
-    """Describes a patch that is retrieved from a file in the repository.
-
-    Arguments:
-        pkg (str): the class object for the package that owns the patch
-        relative_path (str): path to patch, relative to the repository
-            directory for a package.
-        level (int): level to pass to patch command
-        working_dir (str): path within the source directory where patch
-            should be applied
-    """
-
-    def __init__(self, pkg, relative_path, level, working_dir, ordering_key=None):
+    """Describes a patch that is retrieved from a file in the repository."""
+
+    def __init__(
+        self,
+        pkg: "spack.package_base.PackageBase",
+        relative_path: str,
+        level: int,
+        working_dir: str,
+        ordering_key: Optional[Tuple[str, int]] = None,
+        sha256: Optional[str] = None,
```

Comment:
This doesn't make much sense, since it seems for this class we compute the sha from the file. If we pass the sha in, then it might be inconsistent with the file. :thinking:

Reply:
This was done in order to ensure that all `Patch` objects have a `sha256` attribute. Previously we did things dynamically (generate it on the fly) which made it impossible to statically type check. There are also a lot of places where mypy gets mad because `sha256` can be `str | None` but it's required to be `str`. We could solve this by adding `assert isinstance(self.sha256, str)` in places where we know it should have been set by now. Let me know how you want to proceed, it's going to be ugly either way.

Reply:
Can’t this be a class level type annotation and not an init parameter? Look at PackageBase, which has types for the dicts added dynamically by directives.